packages/test/src/test-custom-payload-converter.ts

```diff
@@ -56,6 +58,122 @@ if (RUN_INTEGRATION_TESTS) {
     };
     await Promise.all([worker.run(), runAndShutdown()]);
   });
+
+  // use default dataConverter for Client
+  // use payload-converter-returns-undefined for Worker
+  // Workflow should throw when running toPayload on "success" retval
+  test('toPayload throwing results in task retry', async (t) => {
+    const taskQueue = 'toPayload throwing results in task retry';
+    const client = new WorkflowClient();
+    const handle = await client.start(workflows.successString, {
+      taskQueue,
+      workflowId: uuid4(),
+    });
+
+    const logs: string[] = [];
+    const sinks: InjectedSinks<LogSinks> = {
+      logger: {
+        log: {
+          fn(_, message) {
+            logs.push(message);
+          },
+        },
+      },
+    };
+
+    const worker = await Worker.create({
+      ...defaultOptions,
+      taskQueue,
+      sinks,
+      dataConverter: {
+        payloadConverterPath: require.resolve('./payload-converters/payload-converter-returns-undefined'),
+      },
+    });
+
+    const runPromise = worker.run();
+    runPromise.catch((err: any) => {
+      console.error('Caught error while worker was running', err);
+    });
+
+    try {
+      await asyncRetry(
+        async () => {
+          const { history } = await client.service.getWorkflowExecutionHistory({
+            namespace: 'default',
+            execution: { workflowId: handle.workflowId },
+          });
+          if (
+            !history?.events?.some(
+              ({ workflowTaskFailedEventAttributes }) =>
+                workflowTaskFailedEventAttributes?.failure?.message ===
+                'The Payload Converter method TestPayloadConverter.toPayload must return a Payload. Received `undefined` of type `undefined` when trying to convert `success` of type `string`.'
+            )
+          ) {
+            throw new Error('Cannot find workflow task failed event');
+          }
+        },
+        {
+          retries: 60,
+          maxTimeout: 1000,
+        }
+      );
+    } finally {
+      await handle.terminate();
+    }
+    t.pass();
+
+    worker.shutdown();
+    await runPromise;
+  });
+
+  test('toPayload throws from client.start', async (t) => {
+    const connection = new Connection();
+    const client = new WorkflowClient(connection.service, {
+      dataConverter: {
+        payloadConverterPath: require.resolve('./payload-converters/payload-converter-returns-undefined'),
+      },
+    });
+    await t.throwsAsync(
+      client.start(workflows.twoStrings, {
+        taskQueue: 'test',
+        workflowId: uuid4(),
+        args: ['hello', 'world'],
+      }),
+      {
+        instanceOf: ValueError,
+        message:
+          'The Payload Converter method TestPayloadConverter.toPayload must return a Payload. Received `undefined` of type `undefined` when trying to convert `hello` of type `string`.',
+      }
+    );
+  });
+
+  test('fromPayload throws from client.execute', async (t) => {
```

Comment:
I'd just add a clarification that throwing is done when getting the result.
Not critical.